### ISSUE TITLE
SimulateCartEndpoint AJAX request makes add to cart unreliable due to race condition (2066)

### ISSUE THREAD

**Issue by siliconforks** [woocommerce/woocommerce-paypal-payments#1725]
### Describe the Bug

On the single product page, this plugin makes AJAX requests to the URL `/?wc-ajax=ppc-simulate-cart`, which is handled by the class `WooCommerce\PayPalCommerce\Button\Endpoint\SimulateCartEndpoint`.  These AJAX requests have the unfortunate side effect of making the add to cart functionality unreliable as a result of a race condition: the AJAX requests may be writing to the WooCommerce session at the same time as the add to cart request.

### To Reproduce

Because this issue is caused by a race condition, it is difficult to reproduce this reliably.  However, you can make it much easier by cheating a little bit and editing the file `modules/ppcp-button/src/Endpoint/SimulateCartEndpoint.php` to make the race condition worse, so it will not take a lot of attempts to trigger it:

```
	protected function handle_data(): bool {
		// Try to trigger race conditions.
		sleep( 2 );

		$products = $this->products_from_request();
		...
```

I've done this and created a video to demonstrate the issue.  The WordPress site in the video has the WooCommerce and WooCommerce PayPal Payments plugins installed (and no other plugins).  WooCommerce PayPal Payments has been connected using a sandbox account (a live account would work too).   The theme used is Storefront.

### Screenshots

The video is embedded below.  While watching the video, keep an eye on the number of items in the cart in the top right corner.

[race-condition.webm](https://github.com/woocommerce/woocommerce-paypal-payments/assets/1156399/aa50e1f1-cca0-49ad-b89e-226a71da770b)

### Expected Behavior

The add to cart functionality needs to work reliably.

### Actual Behavior

Here is a summary of what happens in the video:

* The first click on the add to cart button works; the number of items in the cart increases to 1.
* The next add to cart click also works; the number of items increases to 2.
* The next add to cart click does not work; the number of items stays at 2.
* The next add to cart click works; the number of items increases to 3.
* The next 3 add to cart clicks also work, bringing the number of items to 6.
* The next add to cart click does not work; the number of items stays at 6.
* The next add to cart click works; the number of items increases to 7.
* The next add to cart click appears to work at first, as the number of items increases to 8.  But then the number of items spontaneously goes back to 7!
* The next add to cart click actually does work; the number of items increases to 8, and stays there.
* The next add to cart click works too; the number of items increases to 9.
* The next add to cart click works too; the number of items increases to 10.
* The next add to cart click increases the number of items to 11, but then once more we see it spontaneously decrease back to 10.

### Environment

 - WordPress Version:  6.3.1
 - WooCommerce Version: 8.1.1
 - Plugin Version: 2.2.2
 - Browser [e.g. Chrome, Safari] and Version: any browser will work
 - Any other plugins installed: none

### Additional Details

<details><summary>System status</summary>

```
### WordPress Environment ###

WordPress address (URL): http://example.test/woocommerce-paypal-payments
Site address (URL): http://example.test/woocommerce-paypal-payments
WC Version: 8.1.1
REST API Version: ✔ 8.1.1
WC Blocks Version: ✔ 10.9.3
Action Scheduler Version: ✔ 3.6.2
Log Directory Writable: ✔
WP Version: 6.3.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.57 (Win64) OpenSSL/3.1.2 PHP/7.4.33
PHP Version: 7.4.33
PHP Post Max Size: 100 MB
PHP Time Limit: 120
PHP Max Input Vars: 1000
cURL Version: 7.70.0
OpenSSL/1.1.1s

SUHOSIN Installed: –
MySQL Version: 5.7.39-log
Max Upload Size: 100 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 8.1.1
WC Database Prefix: wp_
Total Database Size: 5.39MB
Database Data Size: 3.66MB
Database Index Size: 1.73MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 2.48MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 0.11MB + Index: 0.06MB + Engine InnoDB
wp_posts: Data: 0.06MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_orders: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_wc_orders_meta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 24
page: 7
post: 2
product: 18
product_variation: 7
wp_navigation: 1
wp_template: 2

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ✔

### Active Plugins (2) ###

WooCommerce PayPal Payments: by WooCommerce – 2.2.2
WooCommerce: by Automattic – 8.1.1

### Inactive Plugins (0) ###


### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature screen enabled: –
HPOS feature enabled: –
Order datastore: WC_Order_Data_Store_CPT
HPOS data sync enabled: –

### WC Pages ###

Shop base: #7 - /shop/
Cart: #8 - /cart/
Checkout: #9 - /checkout/
My account: #10 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront
Version: 4.5.2
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### WooCommerce PayPal Payments ###

Onboarded: ✔
Shop country code: US
WooCommerce currency supported: ✔
Advanced Card Processing available in country: ✔
Pay Later messaging available in country: ✔
Webhook status: –
PayPal Vault enabled: –
ACDC Vault enabled: –
Logging enabled: –
Reference Transactions: ✔
Used PayPal Checkout plugin: –
Tracking enabled: ✔

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: customize-store
minified-js
new-product-management-experience
product-variation-management
settings
async-product-editor-category-field

Daily Cron: ✔ Next scheduled: 2023-09-23 20:03:40 +00:00
Options: ✔
Notes: 63
Onboarding: -

### Action Scheduler ###

Complete: 38
Oldest: 2023-09-23 05:56:04 +0000
Newest: 2023-09-23 06:04:29 +0000

Pending: 1
Oldest: 2023-09-24 05:56:04 +0000
Newest: 2023-09-24 05:56:04 +0000


### Status report information ###

Generated at: 2023-09-23 07:10:08 +00:00
```

</details>

**Comment by siliconforks** [comment]
Just a brief follow-up on this: I have one web site which is running the WooCommerce PayPal Payments plugin and is also running the [Klaviyo plugin](https://wordpress.org/plugins/klaviyo/) (which basically just collects data and metrics).  After upgrading WooCommerce PayPal Payments, I noticed that the new version appeared to be causing problems with the Klaviyo plugin.  I was going to file another issue about this, but it's almost certainly caused by the same thing as this issue (the `/?wc-ajax=ppc-simulate-cart` AJAX requests handled by the `SimulateCartEndpoint` class), so I'll just mention it here:

![klaviyo](https://github.com/woocommerce/woocommerce-paypal-payments/assets/1156399/f624f393-8ecc-4e06-80a5-21cde224c17b)

On September 9 I updated the WooCommerce PayPal Payments plugin, and Klaviyo started recording obviously incorrect add to cart metrics.  On September 22, I modified the WooCommerce PayPal Payments plugin settings to disable the "PayPal Smart Buttons" and "Pay Later" features, which appears to cause the plugin to no longer make the `/?wc-ajax=ppc-simulate-cart` AJAX requests; the Klaviyo add to cart metrics returned to normal.

In summary: I think these `/?wc-ajax=ppc-simulate-cart` AJAX requests really need to be reconsidered - they seem to be causing a **lot** of issues.

**Comment by InpsydeNiklas** [comment]
Hi @siliconforks, thanks for sharing your findings! 
One way to temporarily work around any issues relating to adding products to the cart from the single product page or potentially increased server load would be to disable the PayPal button on the single product page.
Our team is looking into improving the performance and compatibility impact of the cart simulation, which was added to provide better compatibility with extensions that customize product attributes on the single product page (e.g. [WooCommerce Bookings](https://woocommerce.com/products/woocommerce-bookings/) or [Product Add-Ons](https://woocommerce.com/products/product-add-ons/)).
The latest update 2.3.0+ should improve the cart simulation behavior, but we acknowledge there may still be room to improve, and feedback like yours is greatly appreciated. Can you let us know if the last update has a measurable impact on your end and if disabling the button brings everything back to normal? Thank you!

**Comment by siliconforks** [comment]
In my testing it appears that it is necessary to disable both the PayPal button **and** the "Pay Later" feature - if both of those are disabled, then the plugin will not make any more AJAX `/?wc-ajax=ppc-simulate-cart` requests.  So that's the approach I'm currently using.

I have not tried version 2.3 yet - I will test it and see if it improves anything.

**Comment by rsm0128** [comment]
I'm using v2.3.1 and the problem is still there.

**Comment by InpsydeNiklas** [comment]
@siliconforks the next version should resolve the race condition outlined in this issue and additionally implements filters for you to have more control over the cart simulation behavior:
```
//disable simulate cart
add_filter( 'woocommerce_paypal_payments_simulate_cart_enabled', '__return_false' );

//reduce simulate cart frequency
function my_custom_simulate_cart_throttling( $throttling ) {
    return 10000;  // Return 10000 regardless of the initial $throttling value
}
add_filter( 'woocommerce_paypal_payments_simulate_cart_throttling', 'my_custom_simulate_cart_throttling' );

//prevent shutdown suppression if there are any problems with the cart behavior
add_filter( 'woocommerce_paypal_payments_simulate_cart_prevent_updates', '__return_false' );
```
A [pre-release version](https://github.com/woocommerce/woocommerce-paypal-payments/releases) with these changes should be available for testing within the coming week. I hope this helps!
